include/boost/random/sobol.hpp

```diff
@@ -0,0 +1,1120 @@
+};
+
+} // namespace detail
+/** @endcond */
+
+//!class template sobol implements a quasi-random number generator as described in
+//! \blockquote
+//![Bratley+Fox, TOMS 14, 88 (1988)]
+//!and [Antonov+Saleev, USSR Comput. Maths. Math. Phys. 19, 252 (1980)]
+//! \endblockquote
+//!
+//!\attention \b Important: This implementation supports up to 1111 dimensions.
+//!
+//!In the following documentation @c X denotes the concrete class of the template
+//!sobol returning objects of type @c IntType, u and v are the values of @c X.
+//!
+//!Some member functions may throw exceptions of type @c std::overflow_error. This
+//!happens when the quasi-random domain is exhausted and the generator cannot produce
+//!any more values. The length of the low discrepancy sequence is given by
+//! \f$L=Dimension \times 2^{digits}\f$, where digits = std::numeric_limits<IntType>::digits.
+//!
+//! \copydoc friendfunctions
+template<typename IntType>
+class sobol : public detail::gray_coded_qrng_base<
+                        sobol<IntType>
+                      , detail::sobol_lattice<IntType>
+                     >
+{
+  typedef sobol<IntType> self_t;
+  typedef detail::sobol_lattice<IntType> lattice_t;
+  typedef detail::gray_coded_qrng_base<self_t, lattice_t> base_t;
+
+public:
+  typedef IntType result_type;
+
+  /** @copydoc boost::random::niederreiter_base2::min() */
+  static BOOST_CONSTEXPR result_type min BOOST_PREVENT_MACRO_SUBSTITUTION ()
+  { return 0; }
+
+  /** @copydoc boost::random::niederreiter_base2::max() */
+  static BOOST_CONSTEXPR result_type max BOOST_PREVENT_MACRO_SUBSTITUTION ()
+  { return (std::numeric_limits<IntType>::max)(); }
+
+  //!Effects: Constructs the default s-dimensional Sobol quasi-random number generator.
+  //!
+  //!Throws: bad_alloc, invalid_argument, range_error.
+  explicit sobol(std::size_t s)
+    : base_t(s)
+  {}
+
+  // default copy c-tor is fine
+
+  /** @copydoc boost::random::niederreiter_base2::seed() */
+  void seed()
+  {
+    base_t::reset_state();
+  }
+
+  /** @copydoc boost::random::niederreiter_base2::seed(std::size_t) */
+  void seed(std::size_t init)
+  {
+    base_t::seed(init, "sobol::seed");
+  }
+
+  //=========================Doxygen needs this!==============================
+
+  //!Requirements: *this is mutable.
+  //!
+  //!Returns: Returns a successive element of an s-dimensional
+  //!(s = X::dimension()) vector at each invocation. When all elements are
+  //!exhausted, X::operator() begins anew with the starting element of a
+  //!subsequent s-dimensional vector.
+  //!
+  //!Throws: overflow_error.
+
+  // Fixed in Doxygen 1.7.0 -- id 612458: Fixed problem handling @copydoc for function operators.
+  result_type operator()()
+  {
+    return base_t::operator()();
+  }
+
+  /** @copydoc boost::random::niederreiter_base2::discard(std::size_t) */
+  void discard(std::size_t z)
+  {
+    base_t::discard(z);
+  }
+};
+
+} // namespace random
+
+typedef random::sobol<uint32_t> sobol;
```

Comment:
If generating a point in the Sobol sequence takes ~50ns (haven't done performance testing, just seems to be a reasonable assumption), then shouldn't a 32 bit number overflow in roughly 3-4 minutes, on a single thread? This default seems likes it's too small; a high dimensional numerical integration can last for days. The default typedef should probably be 64 bit, no?

Reply:
I think you are right. If I parametrize Sobol lattice bits as uint32_t, then it naturally gives seq_count just up to 2^32-1 (trivial zeros are skipped).

Perhaps the default should indeed be uint64_t nowadays. I should also probably store seq_count as that IntType, and update seed and discard member fn param types appropriately.